src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/DynamicMethod.cs

```diff
@@ -431,38 +465,145 @@ internal RuntimeMethodHandle GetMethodDescriptor()
-                parameters![index] = arguments[index];
+        // Slower path that does a heap alloc for copyOfParameters and registers byrefs to those objects.
+        // This is a separate method to support better performance for the faster paths.
+        private static unsafe object? InvokeWithManyArguments(
+            DynamicMethod mi,
+            int argCount,
+            object? obj,
+            BindingFlags invokeAttr,
+            Binder? binder,
+            object?[]? parameters,
+            CultureInfo? culture)
+        {
+            Debug.Assert(parameters != null);
+
+            object[] objHolder = new object[argCount];
+            Span<object?> copyOfParameters = new Span<object?>(objHolder, 0, argCount);
+
+            // We don't check a max stack size since we are invoking a method which
+            // naturally requires a stack size that is dependent on the arg count\size.
+            IntPtr* pByRefStorage = stackalloc IntPtr[argCount];
+            Buffer.ZeroMemory((byte*)pByRefStorage, (uint)(argCount * sizeof(IntPtr)));
+
+            bool* boolHolder = stackalloc bool[argCount];
+            Span<bool> shouldCopyBackParameters = new Span<bool>(boolHolder, argCount);
+
+            GCFrameRegistration reg = new(pByRefStorage, (uint)argCount, areByRefs: true);
+
+            object? retValue;
+            try
+            {
+                RegisterForGCReporting(&reg);
+                mi.CheckArguments(
+                    copyOfParameters,
+                    pByRefStorage,
+                    shouldCopyBackParameters,
+                    parameters,
+                    mi.Signature.Arguments,
+                    binder,
+                    culture,
+                    invokeAttr);
+
+                retValue = mi.Invoker.InvokeUnsafe(obj, pByRefStorage, invokeAttr);
+            }
+            finally
+            {
+                UnregisterForGCReporting(&reg);
+            }
+
+            // Copy modified values out. This should be done only with ByRef or Type.Missing parameters.
+            for (int i = 0; i < argCount; i++)
+            {
+                if (shouldCopyBackParameters[i])
+                {
+                    parameters[i] = copyOfParameters[i];
+                }
+            }
 
-            GC.KeepAlive(this);
             return retValue;
         }
 
+        [DebuggerHidden]
+        [DebuggerStepThrough]
+        internal unsafe object? InvokeNonEmitUnsafe(object? obj, IntPtr* arguments, BindingFlags invokeAttr)
+        {
+            if ((invokeAttr & BindingFlags.DoNotWrapExceptions) == 0)
+            {
+                try
+                {
+                    return RuntimeMethodHandle.InvokeMethod(obj, (void**)arguments, Signature, isConstructor: false);
+                }
+                catch (EntryPointNotFoundException)
+                {
+                    // Don't wrap since the exception did not originate from within the method.
```

Comment:
This is not correct. EntryPointNotFoundException can be thrown by any PInvoke and thus can definitely originate from within the method.

What is the corner case that this is trying to handle?

Reply:
Yes thanks for looking at this. The prior exception semantic changes this broke at least one test assuming `EntryPointNotFoundException`; originally I thought the exception was coming from within `::InvokeMethod`. I pushed a new commit to revert that and update the test, although I believe there will still be a test failure elsewhere that is likely related.

Also we need to evaluate the Mono code in this area since the managed invoke exception handler re-throws `MethodAccessException`, `OverflowException` and `Mono.NullByRefReturnException` (which re-throws as `NullReferenceException`). Catching and re-throwing these may not be correct in all cases either.